presto-spi/src/main/java/com/facebook/presto/spi/relation/LogicalRowExpressions.java

```diff
@@ -261,6 +261,55 @@ public RowExpression pushNegationToLeaves(RowExpression expression)
         return expression.accept(new PushNegationVisitor(), null);
     }
 
+    /**
+     * Given a logical expression, the goal is to convert to conjuctive normative form (CNF).
+     * This requires making a call to `pushNegationToLeaves`.
+     * This only applies to propositional logic. this utility cannot be applied to high-order logic.
+     * Examples of non-applicable cases could be f(a AND b) > 5
+     *
+     * An applicable example:
+     *
+     *        NOT
+     *         |
+     *      ___OR_                          AND
+     *     /      \                      /      \
+     *    NOT     OR        ==>        OR      AND
+     *     |     /  \                 /  \     /   \
+     *    OR   c   NOT              a    b   NOT  d
+     *   /  \        |                         |
+     *  a    b       d                         c
+     */
+    public RowExpression convertToConjunctiveNormativeForm(RowExpression expression)
+    {
+        return convertToNormativeForm(expression, AND);
+    }
+
+    /**
+     * Given a logical expression, the goal is to convert to disjunctive normative form (DNF).
+     * The same limitations apply as for conjunctive normative form (CNF).
+     *
+     * An applicable example:
+     *
+     *        NOT                                    OR
+     *         |                                 /        \
+     *      ___OR_                          AND            AND
+     *     /      \                        /    \        /     \
+     *    NOT     OR        ==>          a     AND      b     AND
+     *     |     /  \                         /   \          /   \
+     *    OR   c   NOT                       NOT  d         NOT  d
+     *   /  \        |                         |              |
+     *  a    b       d                         c              c
+     */
+    public RowExpression convertToDisjunctiveNormativeForm(RowExpression expression)
+    {
+        return convertToNormativeForm(expression, OR);
+    }
+
+    public RowExpression convertToNormativeForm(RowExpression expression, Form clauseJoiner)
```

Comment:
Could you add javadoc to warn the risk of using this interface? Also, it would be nice to have a hard stop. For example, have a mechanism to check if the given expression is too large. cc: @wenleix

Reply:
Sure and I agree. I'm thinking of having a 2-fold threshold - 1 on the # of total terms (to prevent too many filters) and 1 on the relative expansion (to skip pushdown if applying these filters may be too slow). What are your thoughts on this @wenleix ?

Reply:
I'm ok to have the util without a threshold for this patch. A javadoc could be good enough.

Reply:
Sure thing, I'll update the javadoc